MennaMohie/assignment2/Solution.cpp

```diff
@@ -0,0 +1,174 @@
+#include<bits/stdc++.h>
+using namespace std;
+
+template<class T>
+class BinaryTree;
+
+template<class T>
+class Node
+{
+    T value;
+    Node *left, *right;
+    friend class BinaryTree<T>;
+
+public:
+    Node():left(nullptr),right(nullptr) {}
+    Node(T val):left(nullptr),right(nullptr),value(val) {}
+
+    T getValue(){return value;}
+};
+
+template<class T>
+class BinaryTree
+{
+    Node<T> *root;
+
+public:
+    BinaryTree(T val):root(new Node<T>(val)) {}
+
+    Node<T>* getRoot()
+    {
+        return root;
+    }
+
+    bool contains(Node<T> *currentRoot, T val)
+    {
+        if(currentRoot == nullptr)
+            return false;
+        if(currentRoot->value == val)
+            return true;
+        return (contains(currentRoot->left, val) || contains(currentRoot->right, val));
+    }
+
+    bool resetRoot(T val)
+    {
+        if(root==nullptr)
+        {
+            root=new Node<T>(val);
+            return true;
+        }
+        else if(!contains(root, val))
+        {
+            root->value=val;
+            return true;
+        }
+        else
+            return false;
+    }
+
+    //adds a new node (val) as a left child to an existing node (parentVal)
+    bool addAsLeftChild(Node<T> *currentRoot, T parentVal, T val)
+    {
+        if(!contains(root, parentVal) || contains(root, val) || currentRoot==nullptr)
+            return false;
+        if(currentRoot->value==parentVal)
+        {
+            if(currentRoot->left==nullptr)
+            {
+                currentRoot->left=new Node<T>(val);
+                return true;
+            }
+            else
+            {
+                return false;
+            }
+        }
+        if(addAsLeftChild(currentRoot->left, parentVal, val))
+            return true;
+        return addAsLeftChild(currentRoot->right, parentVal, val);
+
+    }
+
+    bool addAsRightChild(Node<T> *currentRoot, T parentVal, T val)
```

Comment:
If possible, it will be nice to merge the addAsLeftChild & addAsRightChild functions into one (with a extra parameter to differentiate). This extra parameter can be a bool, or a enum (preferred).